h2o-core/src/main/java/water/FailedNodeWatchdogExtension.java

```diff
@@ -0,0 +1,155 @@
+        return "Failed node watchdog";
+    }
+
+    @Override
+    public void printHelp() {
+        System.out.println(
+                "\nFailed node watchdog extension:\n" +
+                        "    -watchdog_client_retry_timeout\n" +
+                        "          Time in milliseconds specifying in which intervals the failed nodes are checked. If not \n" +
+                        "          specified, the default value of 6000 ms is used. \n" +
+                        "    -watchdog_client\n" +
+                        "          Same as the client except the that cluster is stopped when this client \n" +
+                        "          disconnects from the rest of the cloud or the cloud is stopped when it doesn't \n" +
+                        "          hear heartbeat from the client for specified amount of time."
+        );
+    }
+
+    @Override
+    public String[] parseArguments(String[] args) {
+        for (int i = 0; i < args.length - 1; i++) {
+            H2O.OptString s = new H2O.OptString(args[i]);
+            if (s.matches("watchdog_client_retry_timeout")) {
+                watchdogClientRetryTimeout = s.parseInt(args[i + 1]);
+                String[] new_args = new String[args.length - 2];
+                System.arraycopy(args, 0, new_args, 0, i);
+                System.arraycopy(args, i + 2, new_args, i, args.length - (i + 2));
+                return new_args;
+            }
+            else if(s.matches("watchdog_client")){
+                watchDogClient = true; H2O.ARGS.client = true;
+                String[] new_args = new String[args.length - 1];
+                System.arraycopy(args, 0, new_args, 0, i);
+                System.arraycopy(args, i + 1, new_args, i, args.length - (i + 1));
+                return new_args;
+            }
+        }
+        return args;
+    }
+
+
+    public void validateArguments() {
+        if (watchdogClientRetryTimeout < 0) {
+            H2O.parseFailed("Watchdog client retry timeout has to be positive: " + watchdogClientRetryTimeout);
+        }
+    }
+
+    @Override
+    public void onLocalNodeStarted() {
+        new FailedNodeWatchdogThread().start();
+        H2O.SELF._heartbeat._watchdog_client = watchDogClient;
+    }
+
+
+    /**
+     * This method checks whether the client is disconnected from this node due to some problem such as client or network
+     * is unreachable.
+     */
+    private static void handleClientDisconnect(H2ONode node) {
+        if(node._heartbeat._watchdog_client){
+            Log.warn("Watchdog client " + node + " disconnected!");
+            BullyClientGoneTask tsk = new BullyClientGoneTask(node);
+            Log.warn("Asking the rest of the nodes in the cloud whether watchdog client is really gone.");
+            if(((BullyClientGoneTask)tsk.doAllNodes()).consensus) {
+                Log.fatal("Stopping H2O cloud since the watchdog client is disconnected from all nodes in the cluster!");
```

Comment:
What happens if say only one node sees the client? The bully task would come back with consensus == false and would be send repeatedly? Not a big deal but small things add up. I would maybe add it to heartbeat instead. I.e. last contact with the master client. Or add a increasing time period to prevent this scenario.